containerd-shim-spin-config.md

```diff
@@ -0,0 +1,235 @@
+
+```sh
+systemctl restart containerd
+```
+
+> note, for k3s, edit the k3s service (`systemctl edit k3s`)
+
+## Proposal
+
+As laid out in the [background](#background), this SKIP aims to enable the following categories of configuration:
+
+1. Spin execution configuration
+2. Shim execution configuration
+3. Shim execution environment variables
+
+The first category discusses how to configure an app's execution, while the later two discuss how to configure the platform.
+
+### User Interface
+
+How does a user set this configuration? As is standard with Kubernetes, configuration should be set in Custom Resource Definitions. Rather than creating a new CRD for configuration, the `SpinExecutor` and `SpinApp` CRDs can be extended to support configuring the platform and app execution, respectively. Users should be able to configure Spin execution next to their app configuration in the `SpinApp` CRD. On the other hand, platform configuration is not done on a per app basis. It could be configured in the `SpinAppExecutor` CRD. This would help users easily see how the platform is configured for the executor their app is using. It would also enable creating multiple executors with distinct configuration.
+
+### Configuring Spin Execution in Container Environment Variables
+
+All Spin execution configuration will be defined as known environment variables with an expected range of values. All trigger configuration keys have the prefix "SPIN_<TRIGGER NAME>". Each executor will define which environment variables it supports. The following is a subset of configuration values that may be supported:
+
+| Key | Spin CLI |  Example Value|
+| ---- | ---- | ---- |
+| SPIN_HTTP_LISTEN_ADDR | `spin up --listen` | "0.0.0.0:3000" |
+| OTEL_EXPORTER_OTLP_ENDPOINT | `OTEL_EXPORTER_OTLP_ENDPOINT=http://123.4.5.6:4318" spin up` | "http://123.4.5.6:4318" |
+| SPIN_LOG_DIR | `spin up --log-dir /tmp/log` | "/tmp/log" |
+| RUNTIME_CONFIG_FILE | `spin up --runtime-config-file /var/config.toml` | "/var/config.toml" |
+
+These values will be configured as environment variables in the Container spec. At first, this feels unadvisable, since it differs from user's previous understanding of Linux container environment variables. Deploying a Spin app shouldn't be compared to deploying a Linux container. A Linux container has one virtualized environment and listener. In a Linux container context, environment variables are set in the app execution and affect the business logic. On the other hand, a Spin app can have N many triggers and components, each of which has its own virtualized environment. While Spin does support setting environment variables in components, they are static. We have never encouraged the use of `spin up --env`, as it is more of a developer shortcut, since it sets the same set of environment variables in all components of an app. This makes each sandbox bigger than necessary and doesn't emphasize the component model's strengths.
```

Comment:
What is encouraged instead of `spin up --env`?

Reply:
To use configuration variables.

Reply:
where are those configuration variables set?

Reply:
They can be set in the SpinApp CRD: https://www.spinkube.dev/docs/spin-operator/reference/spin-app/#spinappspecvariablesindex

But all they are is the equivalent of a container env var that is prefixed with `SPIN_VARIABLE_`. [This](https://github.com/spinkube/spin-operator/blob/30e2d8f6ae0261f4c45668f45aa457e557b37c33/internal/controller/deployment.go#L117) is where the Spin Operator does the injection into the container spec. Here are the Spin docs on using the environment variable provider for configuring configuration variables: https://github.com/spinkube/spin-operator/blob/30e2d8f6ae0261f4c45668f45aa457e557b37c33/internal/controller/deployment.go#L117

Reply:
Thanks, that helps my understanding a lot. It is the Spin host that is responsible for lifting the env's into the wasm instance based on the [application variables](https://developer.fermyon.com/spin/v2/variables) that are configured.

Reply:
And using runtime config, that provider could be Vault or Azure Key Vault. I updated the doc to clarify this too

Reply:
I talked to @radu-matei about what the default experience should be, and we came to the conclusion that not forwarding the environment variables to components may be too restrictive. This is SpinKUBE and we can expect that users expect to be able to access the Kubernetes injected environment variables as they normally would in a container. All this being said, i updated the document to recommend setting the env vars rather than using application variables

Reply:
Thanks for the patience explain that all to me. I think we flipped on this since your last comment, and I like the current approach which solve the problem but in a wasm based way where they are selectively injected at the users request.

Marking this resolved